extension/js/common/core/crypto/pgp/openpgp-key.ts

```diff
@@ -618,178 +607,226 @@ export class OpenPGPKey {
     }
     throw new Error('No valid signature found in key');
   };
 
-  private static extractExternalLibraryObjFromKey = (pubkey: Key) => {
-    if (pubkey.family !== 'openpgp') {
-      throw new UnexpectedKeyTypeError(`Key type is ${pubkey.family}, expecting OpenPGP`);
+  private static extractExternalLibraryObjFromKey = async (key: Key): Promise<OpenPGP.Key> => {
+    if (key.family !== 'openpgp') {
+      throw new UnexpectedKeyTypeError(`Key type is ${key.family}, expecting OpenPGP`);
     }
-    const opgpKey = (pubkey as unknown as { internal: OpenPGP.key.Key }).internal;
-    if (!opgpKey) {
+    const keyWithPrivateFields = key as KeyWithPrivateFields;
+    const internal = keyWithPrivateFields.internal;
+    if (!internal) {
       throw new Error('Object has type == "openpgp" but no internal key.');
+    } else if (typeof internal === 'string') {
+      keyWithPrivateFields.internal = await opgp.readKey({ binaryKey: Buf.fromRawBytesStr(internal) });
+      return keyWithPrivateFields.internal;
     }
-    return opgpKey;
+    return internal;
   };
 
-  private static extractStrengthUncheckedExternalLibraryObjFromKey = (pubkey: Key) => {
-    if (pubkey.family !== 'openpgp') {
-      throw new UnexpectedKeyTypeError(`Key type is ${pubkey.family}, expecting OpenPGP`);
+  private static extractStrengthUncheckedExternalLibraryObjFromKey = async (key: Key) => {
+    if (key.family !== 'openpgp') {
+      throw new UnexpectedKeyTypeError(`Key type is ${key.family}, expecting OpenPGP`);
+    }
+    const keyWithPrivateFields = key as KeyWithPrivateFields;
+    if (!keyWithPrivateFields.rawKey) {
+      keyWithPrivateFields.rawKey = await opgp.readKey({ armoredKey: keyWithPrivateFields.rawArmored });
     }
-    const raw = pubkey as unknown as { rawKey: OpenPGP.key.Key };
-    return raw?.rawKey;
+    return keyWithPrivateFields.rawKey;
   };
 
-  private static getKeyIgnoringExpiration = async (
-    getter: (keyid?: OpenPGP.Keyid | null, date?: Date, userId?: OpenPGP.UserId | null) => Promise<OpenPGP.key.Key | OpenPGP.key.SubKey | null>,
-    exp: Date | number | null,
-    expired: () => boolean
-  ): Promise<OpenPGP.key.Key | OpenPGP.key.SubKey | null> => {
-    const firstTry = await Catch.undefinedOnException(getter());
-    if (firstTry) {
-      return firstTry;
+  private static getSigningAndEncryptionKeys = async (key: OpenPGP.Key) => {
+    // encryptionKey and signingKey are obtained with undefined value of the date parameter, which is considered current date
+    const encryptionKey = await Catch.undefinedOnException(key.getEncryptionKey());
+    const signingKey = await Catch.undefinedOnException(key.getSigningKey());
+    const possibleExpirations: number[] = [];
+    const primaryKeyExpiration = OpenPGPKey.getExpirationAsDateOrUndefined(await key.getExpirationTime())?.getTime();
+    if (!encryptionKey || !signingKey) {
+      possibleExpirations.push(
```

Comment:
I reworked this method to make use of `getSigningKey` and `getEncryptionKey` calls from OpenPGP.js instead of trying to replicate OpenPGP.js behaviour when selecting keys, but it turned out to be not so simple. Supplying `date: null` (meaning: ignore date) parameter instead of `date: undefined` (meaning: take current date) it is possible to find a (sub)key for singing/encryption, but it may not be the subkey we're looking for ( 1) the subkey may have `created` property after primary key's expiration is in the `key was never usable` unit test, 2) the subkey may tell us an incorrect expiration date, as there may be another subkey with later expiration that didn't show up because OpenPGP.js sorts subkeys by `created` property)
So, for already expired keys the following algo is implemented:
1. create a list of all the subkey's expiration dates (prior to primary key's expiration) 
(todo: we can make it faster by manually collecting expirations from signatures?)
2. call  `getEncryptionKey`/`getSigningKey`  with dates from the list in descending order until we get a usable key.
If a not-expired encryption key is returned for the current date, then we'll call `getEncryptionKey` with the expiration date of the found key to find a next valid key to figure out the final `expiration` date over all the subkeys.
What do you think, @tomholub ? Is this approach good?

Reply:
Done